tools/benchmarksAnalyzer/src/main/kotlin/org/jetbrains/analyzer/SummaryBenchmarksReport.kt

```diff
@@ -0,0 +1,423 @@
+        addFieldChange(kotlinChanges,"Backend flags", prevCompiler.backend.flags.toString(),
+                       curCompiler.backend.flags.toString())
+        addFieldChange(kotlinChanges,"Kotlin version", prevCompiler.kotlinVersion, curCompiler.kotlinVersion)
+    }
+
+    // Calculate avarage results for bencmarks (each becnhmark can be run several times).
+    fun collectMeanResults(benchmarks: Map<String, List<BenchmarkResult>>): BenchTable {
+        return benchmarks.map {(name, resultsSet) ->
+            val repeatSeq = IntArray(resultsSet.size)
+            var curStatus = BenchmarkResult.Status.PASSED
+            var currentWarmup = -1
+
+            // Collect common becnhmark values and check them.
+            resultsSet.forEachIndexed { index, result ->
+                // If there was at least one failure, summary is marked as failure.
+                if (result.status == BenchmarkResult.Status.FAILED) {
+                    curStatus = result.status
+                }
+                repeatSeq[index] = result.repeat
+                if (currentWarmup != -1)
+                    if (result.warmup != currentWarmup)
+                        println("Check data consistency. Warmup value for benchmark '${result.name}' differs.")
+                currentWarmup = result.warmup
+            }
+
+            repeatSeq.sort()
+            // Check if there are missed loop during running benchmarks.
+            repeatSeq.forEachIndexed { index, element ->
+                if (index != 0)
+                    if ((element - repeatSeq[index - 1]) != 1)
+                        println("Check data consistency. For benchmark '$name' there is no run between ${repeatSeq[index - 1]} and $element.")
+            }
+
+            // Create mean and variance benchmarks result.
+            val scoreMeanVariance = Statistics.computeMeanVariance(resultsSet.map{it.score})
+            val runtimeInUsMeanVariance = Statistics.computeMeanVariance(resultsSet.map{it.runtimeInUs})
+            val meanBench = BenchmarkResult(name, curStatus, scoreMeanVariance.mean,
+                    runtimeInUsMeanVariance.mean, repeatSeq[resultsSet.size - 1],
+                    currentWarmup)
+            val varianceBench = BenchmarkResult(name, curStatus, scoreMeanVariance.variance,
+                    runtimeInUsMeanVariance.variance, repeatSeq[resultsSet.size - 1],
+                    currentWarmup)
+             name to MeanVarianceBench(meanBench, varianceBench)
+        }.toMap()
+    }
+
+    // Print report using render.
+    fun print(renderInstance: Render, onlyChanges: Boolean = false, outputFile: String? = null) {
+        val content = renderInstance.render(onlyChanges)
+        outputFile?.let {
+            writeToFile(outputFile, content)
+        } ?: run {
+            println(content)
+        }
+    }
+
+    // Common interface for printing report in different formats.
+    interface Render {
```

Comment:
Maybe move `Render` and `TextRender` to separate file? It will simplify reasoning about `SummaryBenchmarksReport`. For example, `addedBenchs` could be expressed as `mergedReport.filter { it.value.first == null }.map { it.key }` and be moved to `TextRender` from `SummaryBenchmarksReport`.

Reply:
I don't want to open inner structure of SummaryReport, it doesn't seems right design. At the same time I want possibility to add new forms of output easily. In C++ I can make render classes as friends, I was looking for similar in Kotlin, and the most close is inner class.

About counting with lambdas - we discussed this yesterday, but that adds extra 4 looping over all benchmarks, that doesn't seem to be a great idea too. I changed to lamdas everywhere where it doesn't influence on operations number a lot.

Reply:
It will help to decouple two separate processes (analysis of results and rendering of the report). Explicit data passing them will make code really easier.
You can also make `TextRender` a private class in `SummaryBenchmarksReport.kt` so it won't violate incapsulation.

I think in such tool readability is more important. 4 extra O(n) operations on collection of size 100-1000 is really nothing.

BTW, class friendship in C++ is frequently considered as bad design choice :)